Arisa/i18n article (#201)
remix-run/remix-website: data/posts/remix-i18n.md

---

> title: Remix and i18n

Reviewer: ```suggestion
title: Internationalization with Remix
```

Reviewer: A suggestion from @mjackson

---
title: Remix and i18n

> summary: We'll cover how internationalization (i18n) works with Remix. Together, we'll understand the impact of i18n, fundamental logic and how to manage i18n more efficiently with Remix.

Reviewer: Hey @schabibi1, thoughts on this as an alternative summary?

```suggestion
summary: Learn how to implement internationalization (i18n) in your Remix project. Discover the significance of i18n, its underlying principles, and strategies for effective management using Remix.
```

Author: Applied👍

authors:
  - Arisa Fukuzaki
---

Many experts are discussing how to make the web better. Accessibility, UI UX, web performance, you name it. You might not hear about internationalization (i18n) as much as other topics, but it is still essential to improve the web. In this article, we'll cover how i18n works with Remix, understand the impact of i18n, fundamental logic and how to manage i18n more efficiently with Remix.

I also talked about i18n with Remix in my talk at Remix Conf 2023. If youwant to watch the video recording, you can find [my i18n talk here][remixconf-arisa].

<iframe style="width:100%;aspect-ratio:16/9;" src="https://www.youtube.com/embed/IAMKH8oOLes?si=RVXWYiStu-DQZbS8" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

## What is i18n?

i18n stands for internationalization. Between the first character, “i,” and the last character, “n,” from there are 18 characters. In short, i18n is about implementing the structures and features for your applications to be ready to localize content.

> ## Why we should care about i18n?

Author: Agree!

You can find what you read above in this statistics [from Statista][lang-statistics-on-the-internet].

## i18n fundamentals

Before we dive into how i18n works with Remix, let's take a look at the fundamentals of i18n. There are three ways to determine languages and regions in i18n: location of the IP address, [Accept-Language header][accept-lang-mdn] or [Navigator.languages][navigator-langs-mdn], and identifiers in URLs.

### Location of the IP address

This approach is based on the location of the IP address. There are more accurate ways to determine languages and regions. This approach won't provide the best UX for users. For example, if you're traveling to another country, you will see the content in the language of your current country.

### Accept-Language header or Navigator.languages

> This approach is based on the language settings of the browser. It's more accurate than using the location of the IP address. However, this approach provides language information but only provides regional information.

Author: Sure! What do you think about this rephrased version?
"This approach is based on the language settings of the browser. It's more accurate than using the location of the IP address. This approach provides the user's preferred language information, but users cannot switch languages from the UI."

> This approach is based on the language settings of the browser. It's more accurate than using the location of the IP address. However, this approach provides language information but only provides regional information.

Reviewer: That's perfect!

> This approach is based on the language settings of the browser. It's more accurate than using the location of the IP address. However, this approach provides language information but only provides regional information.

Reviewer: c.c. @jacob-ebey if you had any other thoughts here

Also, it's possible to detect if `params.lang` is not a valid language value by creating a reusable helper function. It's a great way to provide the best UX for users.

Let's look more in-depth at what I mentioned with the example code. You can fork and clone this [Optional Segments example app repo][optional-segments-example-repo] to test it out on your local machine.

![A GIF of Remix Optional Segments example app][optional-segments-app-gif]

This example app is partially based on the Remix contact app tutorial. It has a `contacts` route and `contacts` route has a `contactId` parameter.

```
 app/
├── components/
│   └── Header.tsx // language switcher
├── routes/
│   ├── _index.tsx
│   ├── ($lang).contacts.$contactId.tsx
│   └── ($lang).contacts.tsx
├── root.tsx
├── data.tsx // contacts data
└── utils.ts // reusable getLang function to check valid params.lang
```

A `lang` parameter is where you can see it catches all slugs in different nested levels, such as `ja/contacts` and `ja/contacts/ryan-florance` in this example app. It covers the case if you want to implement localized sub-directories without CMSs.

> Built-in parameters like `params.lang` saves you time when implementing i18n-supported routes. To enable Optional Segments, you can add `($lang)` in the routes like this, `app/routes/($lang).contacts.$contactId` to catch `lang` parameter in the route.

Reviewer: This might be a good place to add some simple examples. Something like

> The urls `/ja/contacts/ryan-florence` and `/contacts/ryan-florence` will both match the `app/routes/($lang).contacts.$contactId` route, since `($lang)` is optional. In this example if there is no `lang` param provided, we default to English (`en`).

Author: I'll add this part and will push👍

> Configuring localized sub-directories with functionally working routes is the most important yet time-consufnig part of the i18n implementation process depending on the built-in features from frameworks.

Reviewer: ```suggestion
Configuring localized sub-directories with functionally working routes is the most important yet time-consuming part of the i18n implementation process depending on the built-in features from frameworks.
```

Remix removes the pain by providing helpful parameters, a flexible structure, and the best DX for developers.

> Make sure you follow the [Google SEO guideline][google-seo-guideline] for non-ASCII characters in the URL. It's not recommended to use non-ASCII characters in the URL. (i.e., `ja/contacts/マイケル-ジャクソン`) It's better to use ASCII characters in the URL. (i.e., `ja/contacts/michael-jackson`)

The Optional Segments example app repo also includes a reusable function to check if `params.lang` is not a valid language code. For better UX, detecting when users access the URL with an invalid language code is essential. It is also important to show them that the page with an invalid language slug doesn't exist.

```js
// utils.ts
import { Params } from "@remix-run/react";

export function getLang(params: Params<string>) {
  const lang = params.lang ?? "en";
  if (lang !== "ja" && lang !== "en") {
    throw new Error(`Invalid language ${lang}`);
  }
  return lang;
}

> ```

Reviewer: It would probably be good to update this code and the example to throw a 404 response, since a general error will be a 500 I believe

[Here is an example in the docs](https://remix.run/docs/en/main/guides/not-found#how-to-send-a-404)

Reviewer: Unless of course the 404 is handled elsewhere, in which case it might be good to point that out

Author: Added 404 response & [example app repo](https://github.com/schabibi1/remix-optional-segments/commit/6508bd1ec4121db74c5b8010329034390aa389a1) is updated👍

I'll update the code in the article and push the change.

> As this `getLang` function is also designed to return selected valid `params.lang` values, it can be used to get only the necessary data for the selected language. The contact data object with types in `data.tsx` looks like this.

Reviewer: ```suggestion
The `getLang` function returns the selected valid `params.lang` value, which means it can be used to get only the necessary data for the selected language. The contact data object with types in `data.tsx` looks like this:
```

```js
// data.tsx
type ContactMutation = {
  id?: string;
  avatar?: string;
  twitter?: string;
  notes?: string;
  favorite?: boolean;
  details?: {
    en?: {
      first?: string;
      last?: string;
    },
    ja?: {
      first?: string;
      last?: string;
    }
  }
};
```

> What we aim to receive necessary data would look like this.

Reviewer: ```suggestion
Here is an example of the i18n data we will receive
```

Reviewer: I would maybe put this and the 2 JS examples below the `loader` example, since the loader is what returns this payload

Author: As we synced, changes applied!

```js
// contacts in Japanese for the name list on the sidebar
{
  "contacts": [
    {
      "id": "kent c.-dodds",
      "name": "ケント C. ドッズ"
    },
    {
      "id": "pedro-cattori",
      "name": "ペドロ カトーリ"
    },
    //...
  ]
}
```

```js
// single contact in Japanese for selected contact
{
  "avatar": "https://sessionize.com/image/820b-400o400o2-Ja1KDrBAu5NzYTPLSC3GW8.jpg",
  "twitter": "@BrooksLybrand",
  "name": "ブルックス リーブランド"
}
```

> How we slim down the data with the `getLang` function looks like this.

Reviewer: ```suggestion
We can slim down the data with the `getLang` function to return only the correct translation of the content
```

export const loader = async ({ params }: LoaderFunctionArgs) => {
  const fullContact = await getContacts();
  const lang = getLang(params);

  const contacts = fullContact.map((contact) => ({
    // different properties to get compared to a single contact
    id: contact.id,
    name: `${contact.details?.[lang]?.first} ${contact.details?.[lang]?.last}`,
  }));

  return json({ contacts, lang });
};

export default function ContactsLayout() {
  const { contacts, lang } = useLoaderData<typeof loader>();

  return (
    <>
      <Header />
      <div id="wrapper">
        <div id="sidebar">
          <h1>{lang === "ja" ? `Remix コンタクト` : `Remix Contacts`}</h1>
          <nav>
            {contacts.length ? (
              <ul>
                {contacts.map(({ id, name }) => {
                  return (
                    <li key={id}>
                      <Link to={`${id}`}>{name}</Link>
                    </li>
                  );
                })}
              </ul>
            ) : (
              <p>
                <i>No contacts</i>
              </p>
            )}
          </nav>
        </div>
        {/* ... */}
      </div>
    </>
  );
}
```

> To create a language switcher on the header, we can use `getLang` helper function on the `Header` component `useLocation` and `useParams` hooks from Remix.

Reviewer: ```suggestion
To create a language switcher on the header, we can use `getLang` helper function on the `Header` component with the `useLocation` and `useParams` hooks from Remix.
```

[`useLocation`][useLocation-docs] can be used to get the current pathname (object) and replace the pathname with the selected language.

> [`useParams`][useParams-docs] returns an object key and value pairs of the dynamic params from the current location that are matched by the routes. (i.e., `routes/($lang).contacts.$contactId.tsx` matched by `ja/contacts/glenn-reyes`. `params.contactId` is `glenn-reyes`.)

Reviewer: ```suggestion
[`useParams`][useParams-docs] returns an object key and value pairs of the dynamic params from the current location that are matched by the routes. (e.g., `routes/($lang).contacts.$contactId.tsx` matched by `ja/contacts/glenn-reyes` will return `params.contactId` with a value of `glenn-reyes`).
```

```jsx
// components/Header.tsx

> // ...

Reviewer: ```suggestion
```

import { Link, useLocation, useParams } from "@remix-run/react";
import { getLang } from "~/utils";

export default function Header() {
  const { pathname } = useLocation();
  const params = useParams();
  const lang = getLang(params);

  return (
    <div id="header">
      <h1>
        {lang === "ja" ? `Optional Segments デモ` : `Optional Segments Example`}
      </h1>
      <nav>
        {lang === "ja" ? (
          <Link to={pathname.replace(/^\/ja/, "")}>🇺🇸</Link>
        ) : (
          <Link to={`/ja${pathname}`}>🇯🇵</Link>
        )}
      </nav>
    </div>
  );
}
```

In this way, we can create a language switcher on the header to switch the language and replace the pathname with the selected language.

## More examples and sources

> Suppose you're looking for a different example and sources to see how i18n with Remix works. In that case, I recommend you check out the example from [Dilum Sanjaya's interactive Remix routing example][interactive-routing-dilum]. Dilum built an example app to visualize Remix routing, and his example includes an Optional Segments example from the Remix docs.

Reviewer: ```suggestion
If you're looking for a different example and sources to see how i18n with Remix works, I recommend you check out the example from [Dilum Sanjaya's interactive Remix routing example][interactive-routing-dilum]. Dilum built an example app to visualize Remix routing, and his example includes an Optional Segments example from the Remix docs.
```

> I hope you find this article helpful in understanding how i18n works with Remix and how to manage i18n more efficiently with Remix. If you have any questions or feedback, feel free to reach out to me on [Twitter][twitter-arisa]!

Reviewer: ```suggestion
I hope this article helped you understand how i18n works and how to manage i18n more efficiently with Remix. If you have any questions or feedback, feel free to reach out to me on [Twitter][twitter-arisa]!
```
